### PR TITLE
DrivAerML: 10-epoch linear warmup + gc=1.0

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

10-epoch linear warmup combined with gradient clipping (gc=1.0) will improve DrivAerML training stability and achieve a lower `val_primary/surface_rel_l2_pct` than the current baseline of 3.997%.

**Rationale:** Warmup alone causes divergence (PRs #3069, #3070). Adding gc=1.0 as a stability guard (chopper, PR #3137, 5-epoch warmup+gc) appears to address this. A 10-epoch warmup gives the optimizer more time to establish stable gradient statistics before the full LR kicks in, potentially reducing early-training instability compared to 5-epoch warmup. The gc=1.0 guard also prevents cosine LR restart shocks at cycle boundaries. This fills the gap between 5-epoch warmup+gc (PR #3137) and 20-epoch warmup alone (PR #3119, still in-progress).

## Instructions

Train on DrivAerML with 10-epoch linear warmup and gradient clipping gc=1.0. Use the champion 4L/512d/8H + Fourier + cosine T_max=30 config as the base.

```bash
cd target/icml2026 && python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --grad-clip 1.0 \
  --warmup-epochs 10 \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --enable-fourier \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --epochs 999 \
  --wandb-group dm-warmup-gc
```

**Key changes from baseline:**
- Added `--warmup-epochs 10` (baseline: no warmup)
- Added `--grad-clip 1.0` (baseline: no gradient clipping)
- All other settings identical to baseline

**What to report:**
- `val_primary/surface_rel_l2_pct` at best validation epoch
- Whether training was stable (no divergence/spikes)
- Epoch at best checkpoint
- W&B run link

## Baseline

Current best metrics (PR #2898, piccolo):
- `val_primary/surface_rel_l2_pct`: **3.997%** at epoch 467
- W&B run: bht6h42t
- Config: 4L/512d/8H, AdamW lr=5e-4, cosine T_max=30, no-EMA, Fourier features, 360-min budget, SENPAI_MAX_EPOCHS=9999
- External target: <3.71% (AB-UPT, ~500 epochs)

Reproduce baseline:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw \
  --lr 5e-4 \
  --cosine-t-max 30 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 \
  --model-hidden-dim 512 \
  --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 \
  --max-eval-batches 200
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT franky:

## Results

### DrivAerML: 10-epoch linear warmup + gc=1.0

| Metric | Baseline | This experiment | Delta |
|--------|----------|----------------|-------|
| `val_primary/surface_rel_l2_pct` | **3.997%** (epoch 467) | **11.196%** (epoch 80) | +7.199% (worse) |

**W&B run:** `rp02hbxl` (wandb-applied-ai-team/senpai-v1)
**Best epoch:** 80 (of 87 completed before divergence)

### Command used
```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --grad-clip 1.0 --warmup-epochs 10 \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --enable-fourier --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 --no-use-ema \
  --epochs 999 --wandb-group dm-warmup-gc \
  --agent franky --wandb-name "franky/dm-warmup10-gc1"
```

### What happened

**The hypothesis did not work.** Training showed promising early convergence but suffered from escalating catastrophic instability events that ultimately prevented the model from reaching the baseline.

**Convergence trajectory:**
- Warmup phase (epochs 1-10): Smooth, stable LR ramp from ~0 to 5e-4. Val improved steadily from 58.7% → 22.4%. Warmup worked as intended.
- Early post-warmup (epochs 11-50): Best-metric envelope improved well: 20.5% → 18.8% → 17.4% → 16.4% → 14.9% → 12.1% → 11.5%. Oscillations from cosine restarts (T_max=30 steps) were present but recoverable.
- Epoch 50-80: Continued slow improvement to 11.20% best, but with increasingly severe instability spikes.
- Epoch 82+: **Catastrophic divergence.** Grad norms exploded exponentially: 3,712 → 71,040 → 2.1M → 41.5M. Training killed at epoch 87.

**Pattern of escalating instability:**
| Spike | Epoch range | Peak grad norm | Peak val | Recovery |
|-------|-------------|---------------|----------|----------|
| 1 | 21-32 | ~9 | 54% | ~12 epochs |
| 2 | 54-70 | ~126 | 50% | ~16 epochs |
| 3 | 82+ | 41,500,000 | 74% | **No recovery** |

Each instability event was ~10x worse than the previous one, suggesting the cosine warm restarts (T_max=30 steps) combined with gc=1.0 create a feedback loop where the model enters increasingly unstable basins after each restart cycle. The gc=1.0 clips gradients but doesn't prevent the model parameters from drifting into catastrophically unstable regions over many steps.

**Peak memory:** ~6 GB VRAM (single GPU)

### Bug fix note

Fixed a pre-existing `UnboundLocalError` in `main()` where `primary_metric_key` (a top-level function) was shadowed by a local string variable assignment, causing Python to treat it as unassigned-local on the function call at line 1650. Removed the redundant variable and unified on `best_val_primary_metric_name`.

### Suggested follow-ups

1. **Tighter gradient clipping (gc=0.3 or gc=0.5):** gc=1.0 is clearly insufficient to prevent cascade divergence. The baseline's competitor PR #3137 (chopper, 5-epoch warmup + gc) might reveal whether tighter clipping helps.
2. **Lower peak LR with warmup:** Instead of warmup to 5e-4, try warmup to 3e-4 or 2e-4 to reduce the energy of cosine restart shocks.
3. **eta_min > 0 for CosineAnnealingLR:** The current cosine goes to LR=0, which means the model stops learning at troughs but then gets a full-power restart. A non-zero eta_min (e.g., 1e-5) would smooth transitions.
4. **Longer T_max:** T_max=30 steps gives ~13 cosine cycles per epoch, which is extremely aggressive. A longer T_max (e.g., 100 or 200 steps) would reduce the frequency of restart shocks.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3148 because ADVISOR: 10-epoch warmup + gc=1.0 reaches only 11.196% (ep80) then diverges catastrophically at ep87 — gc=1.0 is insufficient to prevent cascade divergence with T_max=30 restarts. Each restart cycle ~10x worse than previous, grad norms reaching 41.5M. Warmup delays but doesn't prevent the fundamental instability. 7.4 pp worse than the 3.833% baseline. Closing..
